src/random/random.c

```diff
@@ -373,11 +590,9 @@ igraph_real_t igraph_rng_get_unif01(igraph_rng_t *rng) {
     const igraph_rng_type_t *type = rng->type;
     if (type->get_real) {
         return type->get_real(rng->state);
-    } else if (type->get) {
-        unsigned long int max = type->max;
-        return type->get(rng->state) / ((double)max + 1);
+    } else {
+        return ldexp(type->get(rng->state), -igraph_rng_bits(rng));
```

Comment:
This means that that if the underlying RNG provides only a few bits, we end up with very bad quality random reals. 

The interface makes it appears as if it were okay to supply only a few bits per generation round. This needs to be improved in some way, either in documentation or by changing the code.

I'm also somewhat bothered that we use only 32 bits for random reals with the default generator, even though `double` has 52 bits.  Within igraph code, we tend to assume that it's fine to move between integer and real representations of whole numbers, but random reals here are clearly worse than random integers.

Reply:
I checked what assembly compilers generate for `ldexp` using Godbolt, and it seems that they always emit a call to the `ldexp` library function. It suspect that `ldexp` is one of those functions which has complex code paths for overflow/underflow detection (note that it sets `errno`) and is therefore much slower than a more naive multiplication approach.

I am thinking of `hypot(x,y)` which is much slower than `sqrt(x*x + y*y)`. 

My preliminary benchmark of random real generation seems to show that random reals are slower to create than random integers, which is surprising considering that the integer code is much more complex.

I'm not fully comfortable with `ldexp` at the moment ... could we come up with something better here? E.g. store `2^-bits` somewhere and just multiply it naively?

Reply:
As a simple experiment, I replaced the `ldexp` code by `return type->get(rng->state) * 0x1.0p-32;` (knowing that the default RNG produces 32 bits). This took the benchmark timing from 2.8 to 2.0 seconds, a very significant gain. I'm wondering if a similar gain could be achieved by storing the `2^-bits` within the `igraph_rng_t` at seeding time and re-using it.

Reply:
> I'm also somewhat bothered that we use only 32 bits for random reals with the default generator, even though double has 52 bits.

This has always been the case, also with previous versions where we simply generated an integer and then divided it with the maximum integer that the RNG can generate. Since RNGs were 32-bit, we never managed to get better results than that.

I'm a bit wary about doing this:

```
        igraph_real_t r = 0.0;
        uint8_t b = 0;
        while (b < 32 /* DBL_MANT_DIG-1 */) {
            r += type->get(rng->state);
            r *= neg_pow2[type->bits];
            b += type->bits;
        }
```

It seems good in theory, and if we could work with floats in infinite precision, then of course this would be accurate. However, I don't know how the floating-point operations work out, especially in the second (and possibly further) iterations when you are adding a large-ish number from `type->get(rng->bits)` to a small-ish number obtained in the previous iteration in `r`. We might be doing extra work for nothing, or we might be introducing bias into the generated number in ways that we don't understand well enough.

Originally I went with the `ldexp` route because it was suggested [on the PCG page](https://www.pcg-random.org/using-pcg-c-basic.html#generating-doubles). If that's slow, I'm okay to use a single round of multiplication with`0x1.0p-32` but I'm not sure about multiple rounds.

Ultimately, we need to decide whether we want to be precise and unbiased (in which case a much more complex method is required that also caters for subnormal numbers, see `random_real()` [here](https://mumble.net/~campbell/tmp/random_real.c)), or we want to be fast and only reasonably precise and unbiased. I think that for the latter case, 32-bit precision is enough for most users, and those who need greater precision should probably roll their own RNG implementation anyway (and then they can get as fancy as they want in `rng->type->get_real()`).

Reply:
> We might be doing extra work for nothing

It is not for nothing: the second iteration runs only when not all 52 bits have been filled. Note that this is 52, as the 53rd is not represented and implicitly treated as 1.

> might be introducing bias into the generated number in ways that we don't understand well enough

Let's not be superstitious and instead understand what the code does: is there a bias or not.

In a single iteration, this is what happens:

 - Draw an integer `k` from `[0, MAX-1]` where MAX is `2^bits`.
 - Compute `k/MAX` to obtain a float in `[0, 1)`.
 
 In the second iteration, we take this number in `[0, 1)` and add it to an integer drawn from the range `[0, MAX-1]`. The result is still in `[0, MAX)`. Dividing by `MAX` still takes us into `[0, 1)`.
 
This is all correct with infinite precision. Now let's think about what can go wrong with finite precision.

We have 52 bits (the leading bit is considered to be 1). The first round filled out the first 32 bits, leaving the remaining 20 as zero. Note that multiplication by negative powers of 2 doesn't really do much: it only changes the exponent as we never get anywhere near [subnormal](https://en.wikipedia.org/wiki/Subnormal_number) territory.  There is no chance for rounding, which is what distinguishes floats from mathematical reals. The more interesting operation is the addition, which does indeed round.

Suppose that our first random integer is exactly `2^32 - 1`, i.e. all the bits are 1. Now what happens if the second is the same? We only have 52 bits to fill, so 64 - 52 = 12 will be "rounded off". But in which direction, up or down? If down, we have a number that is < 1.0. But if up, we get an exact 1.0. I checked this, and it rounded up (both in C++ and Mathematica with machine numbers). 

We don't exactly get any biases, but there is a risk of getting an exact 1.0, which should not be allowed.

I'll think about how to prevent this without affecting performance or making the code too complex. I suppose the easiest solution would be rejection sampling.

As for biases, I see none, and I do not believe that I missed anything.

Finally, something interesting to note: when I looked at how many bits Mathematica's real generator produces, I noticed that it always leaves the very last bit as zero. Possibly, it has to do with this rounding up issue.

Reply:
Alright, I pushed a fix using rejection sampling, which is simple, and obviously won't produce any bias. Note that we restart the generation from scratch of we encounter an exact 1.0. I am confident about this method.

Note that this fix would have been necessary for the `ldexp` method as well. Here's a transcript of my cling (C++ inrterpreter) session checking this:

```
[cling]$ #include <cmath>
[cling]$ uint64_t a = -1;
[cling]$ a
(unsigned long long) 18446744073709551615
[cling]$ ldexp(a, -64)
(double) 1.0000000
[cling]$ ldexp(a, -64) == 1
(bool) true
```

Reply:
Note regarding biases, just in case someone finds this confusing: 

The distribution of representable floating points numbers is not uniform. There are more (twice as many) representable numbers in [0, 1/2) than in [1/2, 1). Thus this method does not generate all representable floating point numbers with equal probability. But we wouldn't want that. We want our number to fall in [0, 1/2) with the same probability as in [1/2, 1), and this method achieves that.

Reply:
There was in fact a bias: the lowest sampling bin would be underrepresented, as explained in the link you provided @ntamas. I pushed another update to use precisely as many bits as needed. This reduces performance somewhat, unfortunately, but eliminates this bias.

Reply:
I managed to convince myself that this is now probably correct (although it took me a while). I disagree with the rounding, though -- I think that the default IEEE floating-point rounding mode is the banker's rounding (i.e. round to nearest, ties to even), which is not supposed to introduce bias. But, this may still potentially end up rounding towards 1.0 so the rejection rule must still be used in that case.

Reply:
I only saw this now. Let's discuss at next meeting.